stats/statsd/common/collector.go

```
@@ -0,0 +1,205 @@
+			c.pushMetrics()
+		case <-ctx.Done():
+			c.pushMetrics()
+			c.finish()
+			return
+		}
+	}
+}
+
+// GetRequiredSystemTags Return the required system sample tags for the specific collector
+func (c *Collector) GetRequiredSystemTags() lib.TagSet {
+	return lib.TagSet{} // no tags are required
+}
+
+// SetRunStatus does nothing in statsd collector
+func (c *Collector) SetRunStatus(status lib.RunStatus) {}
+
+// Collect metrics
+func (c *Collector) Collect(containers []stats.SampleContainer) {
+	var pointSamples []*Sample
+
+	for _, container := range containers {
+		for _, sample := range container.GetSamples() {
+			c.Summary[sample.Metric.Name] = sample.Metric
+			pointSamples = append(pointSamples, generateDataPoint(sample))
+		}
+	}
+
+	if len(pointSamples) > 0 {
+		c.bufferLock.Lock()
+		c.buffer = append(c.buffer, pointSamples...)
+		c.bufferLock.Unlock()
+	}
+}
+
+func (c *Collector) pushMetrics() {
+	c.bufferLock.Lock()
+	if len(c.buffer) == 0 {
+		c.bufferLock.Unlock()
+		return
+	}
+	buffer := c.buffer
+	c.buffer = nil
+	c.bufferLock.Unlock()
+
+	c.Logger.
+		WithField("samples", len(buffer)).
+		Debugf("%s: Pushing metrics to server", c.Type.String())
+
+	if err := c.commit(buffer); err != nil {
+		c.Logger.
+			WithError(err).
+			Errorf("%s: Couldn't commit a batch", c.Type.String())
+	}
+}
+
+func (c *Collector) finish() {
+	if c.Type == Datadog {
+		c.sendSummaryData()
+	}
+	// Close when context is done
+	if err := c.Client.Close(); err != nil {
+		c.Logger.Debugf("%s: Error closing the client, %+v", c.Type.String(), err)
+	}
+}
+
+func (c *Collector) commit(data []*Sample) error {
+	for _, entry := range data {
+		c.dispatch(entry)
+	}
+	return c.Client.Flush()
+}
+
+func (c *Collector) dispatch(entry *Sample) {
+	var tagList []string
+	if c.FilterTags != nil {
+		tagList = c.FilterTags(entry.Data.Tags, entry.Group)
+	}
+
+	switch entry.Type {
+	case stats.Counter:
+		_ = c.Client.Count(entry.Metric, int64(entry.Data.Value), tagList, 1)
```

I don't like all of the silently ignored errors here...